docs/constant_tensors_cache.md

```diff
@@ -0,0 +1,230 @@
+tensors. They will be marked as interested ops. 
+
+The main work of analysis step will be implemented as a 
+[DataFlow Analysis](https://mlir.llvm.org/docs/Tutorials/DataFlowAnalysis/) pass. In the input MLIR graph,
+the constant tensors will appear as outputs of `arith.constant` operations or arguments of the MLIR module 
+entry function marked with constant attributes. The constantness starts its propagation 
+from these tensors to the output tensors of operations that process them. Eventually, these operations
+will form a subgraph, which is named as 'constant subgraph'. Another subgraph, which contains non-constant operations, 
+consumes the outputs of constant subgraph and the parameters to the graph.
+
+Because the constantness information is carried by tensors, the analysis step is on linalg-on-tensor level. 
+The interested ops are most likely `reorder`, `pack` or `broadcast` ops, so the analysis step should be after the 
+layout propagation pass.
+
+### 3.2 Transform step
+Take the following IR as an example (To make the IR short and easy to understand, only the important information is 
+shown):
+```mlir
+module {
+    // %weight0 and %weight1 is Type 1 constant. %weight2 is Type 2 constant.
+    main(%feature0: tensor<*xbf16>, %weight1: tensor<*xbf16>, %weight2: tensor<*xbf16>) 
+            -> %feature3: tensor<*xbf16> attributes {compiletime_const_args_index = [1 : i32], runtime_const_args_index = [2 : i32]} {
+        %weight0 = arith.constant dense<"0x01234567..."> : tensor<*xbf16>
+        %packedWeight0 = tensor.pack(%weight0, ...)
+        %feature1 = linalg.matmul(%feature0, %packedWeight0)
+        %packedWeight1 = tensor.pack(%weight1, ...)
+        %feature2 = linalg.matmul(%feature1, %packedWeight1)
+        %packedWeight2 = tensor.pack(%weight2, ...)
+        %feature3 = linalg.matmul(%feature2, %packedWeight2)
+        return %feature3
+    }
+}
+```
+
+After transformation, there will be three functions in the module, one for compile time folding, one for runtime 
+folding and one for computing. The compile time folding function contains the operations that consume and produce 
+constants of Type 1. The runtime folding function contains the operations that consume and produce 
+constants of Type 2. The computing function will take all folded tensors as inputs. The expected output IR will be like:
+```mlir
+module {
+    compute(%feature0: tensor<*xbf16>, %foldedWeight0: tensor<*xbf16>, %foldedWeight1: tensor<*xbf16>, %foldedWeight2: tensor<*xbf16>) 
+            -> %feature3: tensor<*xbf16> {
+        %feature1 = linalg.matmul(%feature0, %foldedWeight0)
+        %feature2 = linalg.matmul(%feature1, %foldedWeight1)
+        %feature3 = linalg.matmul(%feature2, %foldedWeight2)
+        return %feature3
+    }
+    compiletime_fold(%weight1: tensor<*xbf16>) -> %foldedWeight0, %foldedWeight1: tensor<*xbf16>, tensor<*xbf16> {
```

Comment:
`compiletime_fold` will be a problem because:
1. The kernel binary will contain the whole tensor to be folded, which is too large. If we only want to use `runtime_fold` from it, the binary size is wasted and this is not friendly for kernel cache.
2. For GPU device, we may want to do the folding on CPU. We shouldn't put three functions in the same module to achieve that.

If we want to support direct compile-time folding, I suggest following the direction of section 2.1 to implement it.

Reply:
For the compile time available tensors, the integration can choose to:

- lower them to arith.constant, which is not suggested, 
- put them into the arguments list of the module and mark them as `compiletime_const_args`,
- put them into the arguments list of the module and mark them as `runtime_const_args`.

For the first two choices, they will be folded by `compiletime_fold`, and for the third choice, by `runtime_fold`. There will be no large literal tensors in the kernel for the last two choices.

Reply:
> We shouldn't put three functions in the same module to achieve that.

I'm not clear if we can generate a new module in the pass pipeline. If so, shall we put the `compiletime_fold` in one module, and `runtime_fold` and `compute` in another module?

Reply:
> I'm not clear if we can generate a new module in the pass pipeline. If so, shall we put the `compiletime_fold` in one module, and `runtime_fold` and `compute` in another module?

I think so. But my current thinking is, we can support `compiletime_fold` in the future when there's a demand for this.

Reply:
OK, I will put all folding operations into `runtime_fold`.